### DISCUSSION TITLE
Updating some 2018 code to Ver 3.X.  Not finding example that helps.

### DISCUSSION THREAD

**Discussion by Edward-2** [pymodbus-dev/pymodbus#1505] [Help] [answered]
I'm trying to get a small pyModbus test [script](https://youtu.be/ZlHqEOmvW6M) to run without errors, but I'm not finding example code that is helpful.  Working with 32-input RS485 modbus modules.  Everything works with MinimalModbus, and an Arduino test, but prefer pyModbus for the slaveID selection in the read line.

Test code:

```
from pymodbus.client import ModbusSerialClient
from pymodbus.register_read_message import ReadHoldingRegistersResponse  #Response

client = ModbusSerialClient(method="rtu", port="/dev/ttyS0", stopbits=1, bytesize=8, parity="N", baudrate=9600, timeout=0.3)

connection = client.connect()
print(connection) # ----- displays  True -----

value = client.read_holding_registers(address=128, count=32, unit=1)  #tested unit=0x01 also
 #  was: value = client.read_input_registers(300, 4. unit=0x01)
  
  # -------- problems below with read and registers terms ----
value = ((f'Modbus error: {read}') if read.isError() else read.registers) # Error line 

print(value.registers)
```

The last error was:
`value = ((f'Modbus error: {read}') if read.isError() else read.registers) # Error line
NameError: name 'read' is not defined`

I think this page [https://pymodbus.readthedocs.io/en/latest/source/library/pymodbus.client.html#serial-rs-485-transport](url) that may have helped is gone.

Any syntax advice greatly appreciated.

**Comment by janiversen** [comment]
your code is wrong, the variable read does not exist the result is in value, so the code code should be "if values.isError etc"

**Comment by janiversen** [comment]
The page you refer to have probably shifted to a new name, there are plenty of documentation including the serial client.

**Reply by Edward-2** [reply]
I know my code is wrong, wanting to correct it.

The read and register errors are gone.
The client.connect() = True.
The (value.isError()) = True

There must be a communication setting missing.  The present error is:
```Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received)```

Code is:
```
from pymodbus.client import ModbusSerialClient
#from pymodbus.client.sync import ModbusSerialClient as ModbusClient
from pymodbus.register_read_message import ReadHoldingRegistersResponse  #Response

client = ModbusSerialClient(method="rtu", port="/dev/ttyS0", stopbits=1, bytesize=8, parity="N", baudrate=9600, timeout=0.3)

connection = client.connect()
print(connection) #indicates True when tested

value = client.read_holding_registers(address=128, count=32, unit=1)  #tested unit=0x01 also
if (value.isError()):
    print(value.isError())
    print(value)
```

**Reply by janiversen** [reply, marked as answer]
Your device do not respond (the error message actually says that).

Did you have a look at documentation at all ? or at the API_Changes.rst that we provide.

unit= is wrong, it should be slave=

**Reply by Edward-2** [reply]
> Did you have a look at documentation at all ? or at the API_Changes.rst that we provide.

Thanks so much janiversen.
The slave solved it. 

I have been looking at the documentation and examples (fragments) for a couple days.
The examples I have been reading in pymodbus.readthedocs (V 2.5.3) were the latest version I found showing the full read_holding-registers command [here](https://pymodbus.readthedocs.io/en/v2.5.3/source/example/async_asyncio_client.html#async-asyncio-client-example) These are all showing "rr = await client.read_holding_registers(1, 1, unit=UNIT)"

Versions 3.2.2 or 3.0.2 had no comparable example, that I could find.  I could find no traditional complete functional examples in pymodbus.readthedocs for serial RTU that would have answered this.

I'm going continue to look for the "slave=" reference in 3.2.2.
I found it searching 'slave=' in 3.3.2 here [https://pymodbus.readthedocs.io/en/v3.2.2/source/examples.html#asynchronous-client-basic-calls](url)

all good now.

**Reply by janiversen** [reply]
examples are not documentation! you should look at the client description instead. Just for information 3.2.2 have a number of examples showing read_holding with different transports including serial.

**Comment by janiversen** [comment]
just read here https://pymodbus.readthedocs.io/en/latest/source/library/client.html#modbus-calls

**Reply by janiversen** [reply]
and the example you claim is not there is just here https://pymodbus.readthedocs.io/en/latest/source/examples.html#asynchronous-client-basic-calls
